src/org/traccar/protocol/NavisProtocolDecoder.java

```diff
@@ -272,70 +316,426 @@ private Object processArray(DeviceSession deviceSession, Channel channel, ByteBu
+        List<Position> positions = new LinkedList<>();
+        int count = buf.readUnsignedByte();
+
+        for (int i = 0; i < count; i++) {
+            Position position = parser.parsePosition(deviceSession, buf).getPosition();
+            if (position.getFixTime() != null) {
+                positions.add(position);
+            }
+        }
+
+        ByteBuf response = Unpooled.buffer(6);
+        response.writeCharSequence(flexHeader, StandardCharsets.US_ASCII);
+        response.writeByte(count);
+        sendFlexReply(channel, response);
+
+        if (positions.isEmpty()) {
+            return null;
+        }
+
+        return positions;
+    }
+
+    private Object processFlexNegotiation(Channel channel, ByteBuf buf) {
+        if ((byte) buf.readUnsignedByte() != (byte) 0xB0) {
+            return null;
+        }
+
+        flexProtocolVersion = (byte) buf.readUnsignedByte();
+        flexStructVersion = (byte) buf.readUnsignedByte();
+        if (flexProtocolVersion != (byte) 0x0A && flexProtocolVersion != (byte) 0x14) {
+            return null;
+        }
+        if (flexStructVersion != (byte) 0x0A && flexStructVersion != (byte) 0x14) {
+            return null;
+        }
+
+        flexBitfieldDataSize = buf.readUnsignedByte();
+        if (flexBitfieldDataSize > 122) {
+            return null;
+        }
+        buf.readBytes(flexBitfield, 0, (int) Math.ceil((double) flexBitfieldDataSize / 8));
+
+        flexDataSize = 0;
+        for (int i = 0; i < flexBitfieldDataSize; i++) {
+            if ((flexBitfield[(int) (i / 8)] & (0x80 >> i % 8)) != 0) {
+                flexDataSize += FLEX_FIELDS_SIZES[i];
+            }
+        }
+
+        ByteBuf response = Unpooled.buffer(9);
+        response.writeCharSequence("*<FLEX", StandardCharsets.US_ASCII);
+        response.writeByte(0xB0);
+        response.writeByte(flexProtocolVersion);
+        response.writeByte(flexStructVersion);
+        sendNtcbReply(channel, response);
+
+        return null;
+    }
+
     private Object processHandshake(Channel channel, SocketAddress remoteAddress, ByteBuf buf) {
-        buf.readByte(); // colon symbol
+        buf.skipBytes(1);  // Colon symbol
         if (getDeviceSession(channel, remoteAddress, buf.toString(StandardCharsets.US_ASCII)) != null) {
-            sendReply(channel, Unpooled.copiedBuffer("*<S", StandardCharsets.US_ASCII));
+            sendNtcbReply(channel, Unpooled.copiedBuffer("*<S", StandardCharsets.US_ASCII));
         }
         return null;
     }
 
-    private static short checksum(ByteBuf buf) {
-        short sum = 0;
+    private static byte xorChecksum(ByteBuf buf) {
+        byte sum = 0;
         for (int i = 0; i < buf.readableBytes(); i++) {
             sum ^= buf.getUnsignedByte(i);
         }
         return sum;
     }
 
-    private void sendReply(Channel channel, ByteBuf data) {
+    private static byte src8Checksum(ByteBuf buf) {
```

Comment:
What's src8?

Reply:
Are you sure we don't have standard method for it?

Reply:
It is a typo. Need to sleep more ...
I was misled by the inline implementation of the XOR checksum in the basic decoder version. Suppose the native implementation can now be used. Necessary changes are made.